lambda_functions/va_profile/va_profile_opt_in_out_lambda.py

```diff
@@ -448,6 +493,157 @@ def make_PUT_request(
+            headers={'Content-Type': 'application/json', 'Authorization': 'Bearer {}'.format(encoded_header)},
+        )
+
+        response = conn.getresponse()
+
+        if response.status == 201:
+            logger.info(
+                'Comp and Pen opt-in confirmation SMS notification sent successfully. Response status: %s',
+                {response.status},
+            )
+            return response
+        else:
+            logger.error(
+                'Failed to send Comp and Pen opt-in confirmation SMS notification. Response status: %s',
+                {response.status},
+            )
+
+    except ValueError as ve:
+        logger.exception(
+            'Configuration error while attempting to send Comp and Pen opt-in confirmation SMS notification: %s', {ve}
+        )
+
+    except Exception as e:
+        logger.exception(
+            'An error occurred while attempting to send Comp and Pen opt-in confirmation SMS notification to vaProfileId %s: %s',
+            va_profile_id,
+            e,
+        )
+    return None
+
+
+def save_notification_id_to_cache(va_profile_id: int, notification_id: str, source_date: str):
+    """
+    Update the VAProfileLocalCache table by inserting the notification_id for the given va_profile_id.
+
+    Args:
+        va_profile_id (int): The VA profile ID of the user the notification was sent to.
+        notification_id (str): The notification UUID.
+    """
+    try:
+        with db_connection.cursor() as cursor:
+            # Execute the SQL query with the provided parameters.
+            cursor.execute(OPT_IN_OUT_ADD_NOTIFICATION_ID_QUERY, (notification_id, va_profile_id, source_date))
+
+            db_connection.commit()
+            logger.info(
+                'Successfully updated VAProfileLocalCache with notification_id %s for va_profile_id %s.',
+                notification_id,
+                va_profile_id,
+            )
+
+    except psycopg2.IntegrityError as e:
+        db_connection.rollback()
+        logger.error(
+            'Failed to insert notification_id %s for va_profile_id %s. IntegrityError: %s',
+            notification_id,
+            va_profile_id,
+            str(e),
+        )
+    except Exception as e:
+        db_connection.rollback()
+        logger.error(
+            'An error occurred while attempting to update notification_id for va_profile_id %s: %s',
+            va_profile_id,
+            str(e),
+        )
+    finally:
+        cursor.close()
+
+
+def base64url(source: bytes) -> str:
+    """
+    Encode a byte source to a base64 URL-safe string.
+
+    Args:
+        source (bytes): The byte source to be encoded.
+
+    Returns:
+        str: The base64 URL-safe encoded string.
+    """
+    encoded_source = base64.b64encode(source).decode('utf-8')
+    encoded_source = encoded_source.rstrip('=').replace('+', '-').replace('/', '_')
+    return encoded_source
+
+
+def generate_jwt(service_api_key: str, service_id: str) -> str:
```

Comment:
why aren't we using the JWT library in this method to do all the heavy lifting? It's already been imported.
also, can we use f-strings instead of .format to be consistent across our code?

Reply:
Updated.